n-api: add `napi_detach_arraybuffer`

legendecas opened the pull request:
As ArrayBuffer#detach is an ecma spec operation ([Section 24.1.1.3](https://tc39.es/ecma262/#sec-detacharraybuffer)), it might be good to have it in N-API.

Fixes https://github.com/nodejs/node/issues/29674

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

addaleax commented on doc/api/n-api.md:
````diff
@@ -3149,6 +3149,26 @@ defined in
+```
+
+* `[in] env`: The environment that the API is invoked under.
+* `[in] arraybuffer`: The JavaScript value of arraybuffer to be detached.
````
```suggestion
* `[in] arraybuffer`: The JavaScript ArrayBuffer to be detached.
```

addaleax commented on src/js_native_api_v8.cc:
```diff
@@ -2997,3 +2997,38 @@ napi_status napi_get_instance_data(napi_env env,
 }
+
+napi_status napi_arraybuffer_detach(napi_env env,
+                                     napi_value arraybuffer) {
```
```suggestion
                                    napi_value arraybuffer) {
```

cjihrig commented on src/js_native_api_v8.cc:
```diff
@@ -2997,3 +2997,38 @@ napi_status napi_get_instance_data(napi_env env,
+  }
+
+  v8::Local<v8::ArrayBuffer> it = value.As<v8::ArrayBuffer>();
+  if (!it->IsExternal()) {
```
Just a question: do you think it's worth documenting that the `ArrayBuffer` must be external and detachable?

jasnell commented on doc/api/n-api.md:
```diff
@@ -3149,6 +3149,26 @@ defined in
+
+This API represents the invocation of the ArrayBuffer Detach operation as
+defined in
+[Section 24.1.1.3](https://tc39.es/ecma262/#sec-detacharraybuffer)
```
nit: for consistency, better to move the URL to an index at the end of the page

vsemozhetbyt commented on doc/api/n-api.md:
```diff
@@ -3149,6 +3149,26 @@ defined in
 [Section 7.2.14](https://tc39.github.io/ecma262/#sec-strict-equality-comparison)
 of the ECMAScript Language Specification.
 
+#### napi_arraybuffer_detach
```
Nit:
```suggestion
### napi_arraybuffer_detach
```

mhdawson commented on src/js_native_api_v8.cc:
```diff
@@ -2997,3 +2997,38 @@ napi_status napi_get_instance_data(napi_env env,
+  if (!value->IsArrayBuffer()) {
+    napi_throw_type_error(
+        env,
+        "ERR_NAPI_INVALID_ARRAYBUFFER_ARGS",
```
The errors should be added to https://github.com/nodejs/node/blob/master/lib/internal/errors.js.  And since the string is different, should likely be different erors as well. For example look at lines 1083 through 1086 as an example of other N-API messages.

gabrielschulhof replied:
In `napi_get_arraybuffer_info()` we use
```C++
RETURN_STATUS_IF_FALSE(env, value->IsArrayBuffer(), napi_invalid_arg);
```
We should not throw here. We should return `napi_invalid_arg`.

gabrielschulhof commented:
It looks like several inconsistencies have crept into how we handle errors in N-API 😕 Although it's too late to say this now, I don't believe N-API should itself throw on error. We should leave that option up to the caller.

I believe it is sufficient to return `napi_invalid_arg` in this case.

gabrielschulhof commented on src/js_native_api_v8.cc:
```diff
@@ -2997,3 +2997,38 @@ napi_status napi_get_instance_data(napi_env env,
+  if (!it->IsExternal()) {
+    napi_throw_type_error(
+        env,
+        "ERR_NAPI_INVALID_ARRAYBUFFER_ARGS",
```
If we do decide to throw here and below then the code for the two errors should be different, because two different messages should not be associated with the same code.

gabrielschulhof replied:
Maybe change the code to `ERR_NAPI_ARRAYBUFFER_NOT_DETACHABLE`.

gabrielschulhof commented on src/js_native_api_v8.cc:
```diff
@@ -2997,3 +2997,38 @@ napi_status napi_get_instance_data(napi_env env,
+  if (!it->IsDetachable()) {
+    napi_throw_type_error(
+        env,
+        "ERR_NAPI_INVALID_ARRAYBUFFER_ARGS",
```
If we do decide to throw, maybe change the code to `ERR_NAPI_ARRAYBUFFER_NOT_DETACHABLE`.

gabrielschulhof commented:
In this PR we also have the option of adding one or more new `napi_status` and returning that upon failure. Such a change would not be semver-major.

devsnek commented on doc/api/n-api.md:
```diff
@@ -3149,6 +3149,26 @@ defined in
+
+This API represents the invocation of the ArrayBuffer Detach operation as
+defined in
+[Section 24.1.1.3](https://tc39.es/ecma262/#sec-detacharraybuffer)
```
should probably just say "defined in ecma262", no guarantee of those numbers staying constant.

legendecas replied:
We already have lots of ecma section references in `n-api.md`. IMO it might be helpful to jump between the spec and n-api document since the detailed operation spec would not be listed in `n-api.md`. If there is any permanent link to ecma spec sections would be better. Or we could constantly update the links.

gabrielschulhof commented:
Furthermore, if we do decide to throw, IMO we should return `napi_pending_exception`.

gabrielschulhof commented:
Can we please call the function `napi_detach_arraybuffer`? We've mostly kept an imperative tone throughout the API so far:
`napi_create_arraybuffer`
`napi_get_arraybuffer_info`
`napi_do_this`
`napi_do_that`

legendecas commented:
@gabrielschulhof Is there any suggestion on distinguishing `napi_invalid_args` in the case with different causes?

I noticed that `napi_extended_error_info` has a field `error_message` but we did not utilize it to supply additional information about the error but a description of the napi_status.

legendecas commented:
Just updated the patch to add two new napi_status and prevent from throwing:

- napi_arraybuffer_expected, and
- napi_detachable_arraybuffer_expected

Also since if an internal storage arraybuffer is detachable was not defined in ECMA spec, I've added a note in the document that requiring an external arraybuffer to be detached is an engine-specific behavior.

/cc @gabrielschulhof @mhdawson @cjihrig

addaleax commented on doc/api/n-api.md:
```diff
@@ -3148,6 +3150,31 @@ Returns `napi_ok` if the API succeeded.
+And another engine specific condition is that an ArrayBuffer would be detachable
+only if it is an external ArrayBuffer created by
+[`napi_create_external_arraybuffer`][]. V8 requires the ArrayBuffer being
+external to be detachable.
```
I’m a bit torn on that – I think I *would* keep this restriction here *because* it’s supposed to be engine-agnostic. If other engines do not have this restriction, then we should attempt to emulate V8’s behaviour for that case everywhere.

gabrielschulhof replied:
@addaleax OK, so how about

> An &#96;ArrayBuffer&#96; is non-detachable if it has been detached before. The engine may impose additional conditions on whether an &#96;ArrayBuffer&#96; is detachable. For example, V8 requires that the &#96;ArrayBuffer&#96; be external, that is, created with [&#96;napi_create_external_arraybuffer&#96;][].

legendecas commented:
Just noticed the latest ECMA spec added a new optional parameter `detach_key` to detach op. Should we add a new `napi_detach_arraybuffer_with_key` later or update the PR?

gabrielschulhof commented:
@legendecas got a link?

legendecas commented:
@gabrielschulhof here it is https://www.ecma-international.org/ecma-262/9.0/#sec-isdetachedbuffer. it was appended on ECMA 2018.

gabrielschulhof commented:
@legendecas your link refers to `IsDetachedBuffer` not to a `detach_key`.

legendecas commented:
> @gabrielschulhof  your link refers to `IsDetachedBuffer` not to a `detach_key`.

Ah, sorry. It's [24.1.1.3](https://www.ecma-international.org/ecma-262/9.0/#sec-detacharraybuffer) DetachArrayBuffer ( arrayBuffer [ , key ] ).

gabrielschulhof commented:
@legendecas I think this PR is good as it is. I filed issues for the `_with_key()` version and for `_is_detached_buffer()`.

legendecas commented:
I think the CI failed for unrelated cases.

- `test.sequential/test-http2-large-file`
- `test.parallel/test-child-process-fork-exec-path`
- `test.parallel/test-child-process-pipe-dataflow`

Trott commented:
Landed in 6afed1d

mafintosh commented:
Anyway to get this backported to 10? We need this is our libsodium bindings, to deallocate secure memory safely.

legendecas commented:
@mafintosh it seems very fortunate that v10 maintenance start date has been delayed to 2020-05-19. I'm willing to seek to land this patch on v10.

mafintosh commented:
@legendecas ah massive thanks! really appreciate it. If we can help somehow let me know.
